### PR TITLE
[DOCS]  Create a new page for grok content in scripting docs

### LINKED ISSUE #71404: [DOCS] Move Grok syntax docs to separate reference page
Currently, our Grok syntax docs reside in the [Grok ingest processor docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/grok-processor.html).

Now that runtime fields also support Grok, we should consider moving those docs into a separate reference section.

Relates to #72580

### LINKED ISSUE #73262: [Doc] Runtime fields and Grok - additional example and Grok pickiness
At https://www.elastic.co/guide/en/elasticsearch/reference/master/runtime-examples.html#runtime-examples-grok we are documenting how to use runtime fields in the mapping. Can we please also elaborate that this can be used in a search query. E.g.:
````
GET my-index/_search
{
  "runtime_mappings": {
    "clientip": {
      "type": "keyword",
      "script": {
        "source": """
        String clientip=grok('%{COMMONAPACHELOG}').extract(doc["message"]).value)?.clientip;
if (clientip != null) emit(clientip);
"""
      }
    }
}
}
````

In addition would it be worth to document that accessing tokenized values (neither `keyword` nor `wildcard` but text with fielddata) via `.extract(params["_source"]["message"])` (instead of `.extract(doc["message"])`).

Finally, Grok seems to be more picky on the runtime fields compared to the Kibana Grok debugger - especially on the whitespaces. E.g. 
Raw: `"clientip = 127.0.0.1"`
Kibana Grok debugger: `clientip = %{COMMONAPACHELOG}` - will match
Runtime Grok: `grok("clientip = %{COMMONAPACHELOG}")` - will not match and requires `grok("clientip\\s=\\s%{COMMONAPACHELOG}")` instead.

Ref. https://github.com/elastic/sdh-elasticsearch/issues/4361

### PR DESCRIPTION
**Pull request by lockewritesdocs**
Adds a new page for grok that describes what it is and how to use it in the context of Painless scripts and runtime fields.

Preview link: https://elasticsearch_73118.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/grok.html

Closes #71404

Closes #73262

### CODE REVIEW

File: docs/reference/mapping/runtime.asciidoc [hunk 1]
```diff
@@ -91,7 +91,7 @@ calculates the day of the week based on the value of `timestamp`, and uses
 
 [source,console]
 ----
-PUT my-index/
+PUT my-index-000001/
```
Comment (nik9000) [inline docs/reference/mapping/runtime.asciidoc:94]: We use `my-index` and stuff like that all the time so I don't think its a huge deal. I'm not sure how realistic the `-0000001` suffix is for queries. I think if folks have time based data they'll likely query an alias.

File: docs/reference/scripting/grok-syntax.asciidoc [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+[[grok-syntax]]
+==== Grok patterns
+The {stack} ships with numerous https://github.com/elastic/elasticsearch/blob/master/libs/grok/src/main/resources/patterns/grok-patterns[predefined grok patterns] that simplify working with grok. The syntax for reusing grok patterns
+takes one of the following forms:
+
+[%autowidth]
+|===
+|`%{SYNTAX}` | `%{SYNTAX:SEMANTIC}` |`%{SYNTAX:SEMANTIC:TYPE}`
+|===
```
Comment (nik9000) [inline docs/reference/scripting/grok-syntax.asciidoc:20]: :+1:

File: docs/reference/scripting/grok-syntax.asciidoc [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+like `55.3.244.1`. 
+
+`SEMANTIC`::
+The identifier you give to the piece of text being matched. For  example, `3.44`
+could be the duration of an event, so you might call it `duration`. The string
+`55.3.244.1` might identify the `client` making a request.
+
+`TYPE`::
+The data type you want to cast your named field. `int`, `long`, `double`,
+`float` and `boolean` are supported types.
```
Comment (nik9000) [inline docs/reference/scripting/grok-syntax.asciidoc:35]: It's probably not worth calling out here - but this is different list than we support in runtime fields. It *does* line up with the list of things `emit` can accept though.

File: docs/reference/scripting/grok-syntax.asciidoc [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+For example, if you're working with Apache log data, you can use the
+`%{COMMONAPACHELOG}` syntax, which understands the structure of Apache logs. A
+sample document might look like this:
+
+[source,txt]
+----
+"timestamp":"2020-04-30T14:30:17-05:00","message":"40.135.0.0 - - 
+[30/Apr/2020:14:30:17 -0500] \"GET /images/hm_bg.jpg HTTP/1.0\" 200 24736"
+----
```
Comment (nik9000) [inline docs/reference/scripting/grok-syntax.asciidoc:71]: might be worth leaving a comment in the asciidoc for anyone that wonders why you have the line break. It certainly makes it more readable!

File: docs/reference/scripting/grok-syntax.asciidoc [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+have write access on a cluster where you're testing a script.
+
+TIP: If you need help building grok patterns to match your data, use the 
+{kibana-ref}/xpack-grokdebugger.html[Grok Debugger] tool in {kib}.
+
+For example, if you're working with Apache log data, you can use the
+`%{COMMONAPACHELOG}` syntax, which understands the structure of Apache logs. A
+sample document might look like this:
+
+[source,txt]
```
Comment (nik9000) [inline docs/reference/scripting/grok-syntax.asciidoc:67]: `[source,js]`? Not sure it matters.

**Review by nik9000 — APPROVED** [review]
The thing you call "SEMANTIC" I've been calling "id" or "name" in my head. It looks like that's what the docs used to call it. Do you think it'd be worth renaming it while we're here?

File: docs/reference/scripting/grok-syntax.asciidoc (hunk 4 above)
Comment (lockewritesdocs) [reply docs/reference/scripting/grok-syntax.asciidoc:71]: ++ will add a comment for future contributors 👍

File: docs/reference/scripting/grok-syntax.asciidoc (hunk 5 above)
Comment (lockewritesdocs) [reply docs/reference/scripting/grok-syntax.asciidoc:67]: Good eye -- I'll update that example.

**Pull request merged**